File: PS2/lifecycle.py
#!/bin/python

############
# PS2
# ECON 41800
# 
# R YANG
# 2013
############

## External Libraries

# numerical/matrix library
import numpy

# optimization library
import scipy.optimize

def testSuite():
    #test0()
    test1()
    #test2()
    #test3()
    return 0

# Utility Function test
"""
def test0():
    
    gamma = 3
    nu = 5.
    beta = 1
    B = -1
    T = 3

    u = utility(gamma, nu, beta, B, T)
    mu = marg_utility(gamma, nu, beta, B, T)

    c = [1,1,1]
    l = [2,1,1]
    a = [0,0,0]

    x = []
    x = x + c
    x = x + l
    x = x + a
    x = numpy.array(x)

    print u(x)
    print mu(x)
"""
# Single period test

def test1():

    # Define parameters
    gamma = 3
    nu = 3
    beta = 0.1
    B = -1.

    wmax = 1.
    amin = 0.
    r = 0.

    tauL = 0.
    tauK = 0.

    T = 4
    period = 1

    lifeCycle(gamma, nu, B, beta, wmax, amin, tauL, tauK, r, T, period=1)
    return 0


## Utility Parameters
# gamma
# nu
# beta - annual discount rate
# B

## Wage Parameters
# wmax - max wage (occurs at t=T/2)

## Tax Parameters
# tauL - proportional tax rate on labor
# tauK - proportional tax rate on interest 

## Other Parameters
# amin
# r - annual interest rate
# T - number of time periods
# period - number of years per period (can be decimal)

## Builds Utility function with given parameters
## NOTE: Returns negative utility for convenient use with minimization algorithms

def utility(gamma, nu, beta, B, T):

    print (gamma, nu, beta, B, T)

    # Construct array of discounts
    betas = beta ** numpy.array(xrange(T))


    #consumption = lambda x: (x[0:T]**(1 - gamma)) / (1 - gamma)
    #labor = lambda x: (x[T:2*T]**(1 + nu)) / (1 + nu)

    #u = lambda x: -numpy.dot(betas, consumption(x) + B * labor(x))

    u = lambda x: -numpy.dot(betas, (x[0:T]**(1 - gamma)) / (1 - gamma) + B * (x[T:2*T]**(1 + nu)) / (1 + nu))

    return u

def marg_utility(gamma, nu, beta, B, T):
    
    betas = beta ** numpy.array(xrange(T))

    consumption = lambda x: - betas * x[0:T]**(-gamma)
    labor = lambda x: -B * betas * x[T:(2*T)]**(nu)
    savings = lambda x: numpy.zeros(T)

    mu = lambda x: numpy.concatenate((consumption(x), labor(x), savings(x)))
    return mu

def foo(x):
    print x
    return x

## Finds path of consumption, labor, and savings which maximizes utility

def lifeCycle(gamma, nu, B, beta, wmax, amin, tauL, tauK, r, T, period=1):

    if period != 1:
        ## If period not yearly, adjust interest and discount rates
        # period = 1 -> 1 year per period

        # adjust interest (r)
        r = (1 + r) ** (period) - 1

        # adjust discount factor (beta) 
        # ** double check this **
        beta = (1 + beta) ** (period) - 1
    
    ## Construct wage as function of time
    # w(t) = 1 - (4 / T)(1 - wmax)t + (4 / T^2)(1 - wmax)t^2
    #wage = lambda t: 1 - (4 / T) * (1 - wmax) * t + (4 / (T * T)) * (1 - wmax) * t * t
    wage = numpy.array(range(T))
    wage = 1 - (4/T) * (1 - wmax) * wage + (4 / (T**2)) * (1 - wmax) * (wage ** 2)
    print "WAGE:", wage

    ## Construct utility function (c,l) -> u
    ## utility = lambda cl: (cl[0]**(1-gamma))/(1-gamma) + (B * cl[1]**(1+nu))/(1-gamma)

    ## Note: x variable = [c_1 ... c_T, l_1 ... l_T, A_1 ... A_T] = 3T by 1 matrix
    ## c_t = x[t]
    ## l_t = x[T + t]
    ## A_T = x[2*T + t]

    ## Set up objective function
    # max \sum \beta^i u(c,l)
    obj = utility(gamma, nu, beta, B, T)

    ## Set up jacobian of objective function
    jac = marg_utility(gamma, nu, beta, B, T)

    ## Set up constraints

    # w_t * l_t * (1 - tauL) + A_{t-1} * (1 + r * (1 - tauK)) = c_t + A_t

    constraints = []
    cons = []

    for t in xrange(T):
        if T == 1:
            # No savings in 1 pd case - this is kind of hacky
            f = lambda x: wage[t] * x[T] * (1 - tauL) - x[0] + 0 * (foo(x))
            constraints.append({'type':'eq', 'fun':f})
            cons.append(f)
            break

        if t == 0:
            # In first period, no initial assets
            f = lambda x: wage[t] * x[T] * (1 - tauL) - x[0] - x[2*T]
            constraints.append({'type':'eq', 'fun':f})
            cons.append(f)
            continue

        if t == T-1:
            # In last period, no final assets
            f = lambda x: wage[t] * x[T+t] * (1 - tauL) + x[2*T + t - 1] * (1 + r * (1 - tauK)) - x[t]
            constraints.append({'type':'eq', 'fun':f})
            cons.append(f)
            continue

        if t > 0 and t < T-1:
            f = lambda x: wage[t] * x[T+t] * (1 - tauL) + x[2*T + t - 1] * (1 + r * (1 - tauK)) - x[t] - x[2*T+t]
            constraints.append({'type':'eq', 'fun':f})
            cons.append(f)
            continue

    ## Set up initial guess (all 0)
    x0 = numpy.array([0] * 3 * T)

    ## Set up bounds
    bounds = []

    # minimal consumption, labor
    # c_t, l_t >= 0 for all t

    bounds = bounds + (2*T)*[(0.0,1e10)]

    # minimal capital (max borrowing)
    # A_t >= amin for all t

    bounds = bounds + T*[(amin,1e10)]

    #bounds = []
    #bounds = bounds + (3*T)*[(0,10)]
    print bounds

    ## Select optimization algorithm
    solver = 'SLSQP'

    ## Solve for optimal consumption, labor, and savings
    res = scipy.optimize.fmin_slsqp(obj, x0, fprime = jac, eqcons = cons, f_eqcons = None, bounds=bounds, iter=1000, iprint=2)

    print "RES"
    print res

    return 0

if __name__ == "__main__":
    print "Running Tests"
    testSuite()

File: PS2/lifecycle.R
# External Libraries
library('ipoptr')

ipoptr_solve <- function(gamma, eta, beta, B, wmax, amin, tauL, tauK, r, T, period=1){

  # Adjust for period other than 1 year
  if (period != 1){
    r <- (1 + r) ^ period - 1
    beta <- (1 + beta) ^ period - 1
  }


  # Construct Wage Function
  wage <- function(t) {
    return(1 - (4/T) * (1-wmax) * t + (4/(T*T)) * (1-wmax) * t^2)
  }


  # Objective Function
  eval_f <- function(x) {

    # Utility at each time t
    c <- (x[1:T] ^ (1 - gamma)) / (1 - gamma) # consumption
    l <- B * (x[(T+1):(2*T)] ^ (1 + eta)) / (1 + eta) # labor
    
    u <- c + l

    # Create betas
    betas <- beta ^ c(1:T)

    discounted.u <- sum(betas * u)

    return(-discounted.u)

  }

  # Gradient of Objective Function
  eval_grad_f <- function(x) {
  
    grad = rep(0,3*T)

    for (i in 1:T) {
      # c_t  
      grad[i] = beta^i * (x[i]^(- gamma))

      # l_t
      grad[T+i] = beta^i * B * (x[T+i]^eta)
    }

    return(grad)
  }

  # Constraints
  eval_g <- function(x) {

    g = rep(0,T)
    
    g[1] = x[T+1] * wage(0) * (1 - tauL) - x[1] - x[2*T+1]

    g[T] = x[T+T] * wage(T-1) * (1 - tauL) + x[2*T + T] * (1 + r * (1 - tauK)) - x[T]

    for (i in 2:T-1) {
      g[i] = x[T+i] * wage(i-1) * (1 - tauL) + x[2*T + i] * (1 + r * (1 - tauK)) - x[i] - x[2*T + i] 
    } 

    return(g)

  }

  # Jacobian of Constraint
  eval_jac_g_structure <- rep(list(c(seq(1, 3*T))), T)
  print(eval_jac_g_structure)
  eval_jac_g <- function(x) {
    jac = NULL
    for (t in 1:T){
      tmp = rep(0,3*T)

      # time t constraint depends on c_t, l_t, A_t, A_t-1

      # wage
      tmp[T+t] = wage(t) * (1 - tauL)

      # consumption
      tmp[t] = -1

      # Savings
      if (t != T){
        tmp[2*T+t] = -1
      }

      # Savings from Previous Pd
      if (t != 1){
        tmp[2*T+t-1] = (1 + r * (1 - tauK))
      }

      jac = append(jac, tmp)
    }

    return(jac)
  }

  # Initial Guess
  x0 <- append(rep(1,2*T), rep(0,T))

  # Bounds

  lb <- append(rep(0,2*T), append(rep(amin,T-1), 0))
  ub <- append(rep(Inf, 3*T-1), 0)
  #print(lb)
  #print(ub)
  # Constraint Bounds
  constraint_lb <- rep(0, T)
  constraint_ub <- rep(0, T)

  # Options
  opts <- list("max_iter"=1000)
  #return(0)
  res <- ipoptr(x0=x0,
		eval_f = eval_f,
		eval_grad_f = eval_grad_f,
		eval_g = eval_g,
		eval_jac_g_structure = eval_jac_g_structure,
		eval_jac_g = eval_jac_g,
		lb = lb,
		ub = ub,
		constraint_lb = constraint_lb,
		constraint_ub = constraint_ub,
		opts = opts)

		       
  print(res)
  return(res)
}	

test1 <- function(){
  gamma = 1.5
  eta = 3
  beta = 0.9
  B = -1.

  wmax = 10.
  amin = 0.
  r = 0

  tauL = 0.
  tauK = 0.

  T = 3
  period = 1

  ipoptr_solve(gamma=gamma,
	       eta=eta,
	       beta=beta,
	       B=B,
	       wmax=wmax,
	       amin=amin,
	       r=r,
	       tauL=tauL,
	       tauK=tauK,
	       T=T,
	       period=period)  
}

test0 <- function(){
  T = 3
  gamma = 1.5
  eta = 3
  beta = 0.9
  B = -1

  x = c(5,1,1,1,1,1,0,0,0)
  
  u <- (x[1:T] ^ (1 - gamma)) / (1 - gamma) + B * (x[T+1:2*T] ^ (1 + eta)) / (1 + eta)


  betas <- beta ^ c(1:T)

  discounted.u <- sum(betas * u)

  return(discounted.u)
}

main <- function(){

  # Run tests
  #test0()
  test1()

}

main()